zkml/src/model/llm.rs

```diff
@@ -92,62 +172,196 @@ where
+                obs.observe(unpadded_seq_len - user_len, &trace);
+            }
         }
         Ok(trace)
     }
 }
 
-pub struct LLMTokenizerObserver<T: LLMTokenizer> {
-    input: String,
-    tokenizer: T,
+impl Driver<Element> {
+    pub fn run<E>(
+        &self,
+        input: Vec<Token>,
+        observer: Option<impl Observer<Element>>,
+    ) -> anyhow::Result<InferenceTrace<'_, E, Element>>
+    where
+        E: ExtensionField + Serialize + DeserializeOwned,
+        E::BaseField: Serialize + DeserializeOwned,
+    {
+        let input = Tensor::new(
+            vec![input.len()].into(),
+            input.into_iter().map(|t| t.as_number()).collect::<Vec<_>>(),
+        )
+        .pad_next_power_of_two();
+        self.run_internal::<E>(input, observer)
+    }
+    pub fn context<E, PCS>(&self) -> anyhow::Result<LLMContext<E, PCS>>
+    where
+        E: ExtensionField + Serialize + DeserializeOwned,
+        E::BaseField: Serialize + DeserializeOwned,
+        PCS: PolynomialCommitmentScheme<E>,
+    {
+        let ctx = Context::<E, PCS>::generate(&self.model, None, None)?;
+        Ok(LLMContext {
+            ctx,
+            config: self.config.clone(),
+            max_context: self.max_context,
+        })
+    }
+    pub fn prove<E, PCS>(
+        &self,
+        ctx: &LLMContext<E, PCS>,
+        trace: InferenceTrace<'_, E, Element>,
+    ) -> anyhow::Result<LLMProof<E, PCS>>
+    where
+        E: ExtensionField + Serialize + DeserializeOwned,
+        E::BaseField: Serialize + DeserializeOwned,
+        PCS: PolynomialCommitmentScheme<E>,
+    {
+        let mut tr: BasicTranscript<E> = BasicTranscript::new(b"model");
+        let prover: Prover<'_, E, _, _> = Prover::new(&ctx.ctx, &mut tr);
+        let io = trace.to_verifier_io();
+        let proof = prover.prove(trace).expect("unable to generate proof");
+        Ok(LLMProof { proof, io })
+    }
 }
 
-impl<N: Number, T: LLMTokenizer> Observer<N> for LLMTokenizerObserver<T> {
+impl<E, PCS> LLMContext<E, PCS>
+where
+    E: ExtensionField + Serialize + DeserializeOwned,
+    E::BaseField: Serialize + DeserializeOwned,
+    PCS: PolynomialCommitmentScheme<E>,
+{
+    pub fn verify(&self, proof: LLMProof<E, PCS>, user_input: Vec<Token>) -> anyhow::Result<()>
+    where
+        E: ExtensionField + Serialize + DeserializeOwned + Number,
+        E::BaseField: Serialize + DeserializeOwned,
+        PCS: PolynomialCommitmentScheme<E>,
+    {
+        // 0. check the size of the output
+        let output = proof.io.output[0].clone();
+        let max_len = output.get_shape().numel();
+        // in any case, the output needs to be less than the max context length
+        ensure!(
+            max_len <= self.config.context_length,
+            "output length is greater than the context length"
+        );
+        // either we reached the specific may context length
+        if let Some(max_context) = self.max_context {
```

Comment:
I don't get the rationale of this check: `self.max_context` seems to be set only when constructing the `context`, so it should be independent from a given run of the model. So how could this condition be equivalent to checking that the proven inference reached the max context length?